### PR TITLE
[#12] [KMM] [Backend] As a user, I can see survey's detail

### LINKED ISSUE #12: [KMM] [Backend] As a user, I can see survey's detail
# Why

When a user selects a survey, we will fetch the survey's detail and show them to the user.

# Acceptance Criteria

- Call the endpoint `/api/v1/surveys/<survey_id>`.
- Show error if the API return error.
- Parse the JSON response to an object if it returns success.

[Document](https://documenter.getpostman.com/view/11835486/T1LJjTqj#142b8597-e176-4e74-95a5-cb48ed5115f0)

### PR DESCRIPTION
**Pull request by phongvhd93**
- Close #12 

## What happened 👀

- Add new function to `SurveyRemoteDateSource` for fetching survey's detail by its ID.
- Add `GetSurveyDetailUseCase`
- Add `QuestionApiModel` and `AnswerApiModel`
- Add new property with type `QuestionApiModel` to `SurveyApiModel`
- Add UT
## Insight 📝

Create a use case for fetching Survey's detail by its ID. 

## Proof Of Work 📹

Screenshot 2023-05-17 at 13 30 49

### CODE REVIEW

**Review by markgravity — APPROVED** [review]

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/SurveyRemoteDataSource.kt [hunk 1]
```diff
@@ -9,11 +9,17 @@ import kotlinx.coroutines.flow.Flow
...
 }
 
 class SurveyRemoteDataSourceImpl(private val apiClient: ApiClient) : SurveyRemoteDataSource {
 
     override fun getSurveys(params: GetSurveysQueryParam): Flow<List<SurveyApiModel>> {
         return apiClient.responseBodyWithParams(Endpoint.SURVEYS, HttpMethod.Get, params)
     }
+
+    override fun getSurvey(id: String): Flow<SurveyApiModel> {
+        return apiClient.responseBody(Endpoint.SURVEYS + "/${id}", HttpMethod.Get)
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/SurveyRemoteDataSource.kt]: We can omit the `{}` here.
```suggestion
        return apiClient.responseBody(Endpoint.SURVEYS + "/$id", HttpMethod.Get)
```

File: shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/data.repository/SurveyRepositoryTest.kt [hunk 2]
```diff
@@ -142,4 +142,31 @@ class SurveyRepositoryTest {
...
+        runTest {
+            given(mockSurveyRemoteDataSource)
+                .function(mockSurveyRemoteDataSource::getSurvey)
+                .whenInvokedWith(any())
+                .thenReturn(flow { emit(MockUtil.mockSurveyApiModel) })
+
+            repository.getSurvey("id").test {
+                this.awaitItem() shouldBe MockUtil.mockSurveyApiModel.toSurvey()
+                this.awaitComplete()
+            }
```
Comment (AVI5HEK) [inline shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/data.repository/SurveyRepositoryTest.kt]: Optional:
```suggestion
        runTest {
            given(mockSurveyRemoteDataSource)
                .function(mockSurveyRemoteDataSource::getSurvey)
                .whenInvokedWith(any())
                .thenReturn(flowOf(MockUtil.mockSurveyApiModel))

            repository.getSurvey("id").first() shouldBe MockUtil.mockSurveyApiModel.toSurvey()
        }
```

File: shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/domain.usecase/GetSurveyDetailUseCaseTest.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    fun setUp() {
+        useCase = GetSurveyDetailUseCaseImpl(mockRepository)
+    }
+
+    @Test
+    fun `when getSurvey is called successfully, it returns a survey`() = runTest {
+        given(mockRepository)
+            .function(mockRepository::getSurvey)
+            .whenInvokedWith(any())
+            .thenReturn(flowOf(MockUtil.mockSurveyApiModel.toSurvey()))
```
Comment (AVI5HEK) [inline shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/domain.usecase/GetSurveyDetailUseCaseTest.kt]: We can keep `MockUtil.mockSurveyApiModel.toSurvey()` in a property, e.g.,
```Kotlin
val expectedSurvey = MockUtil.mockSurveyApiModel.toSurvey()
// ...
useCase("survey_id").first() shouldBe expectedSurvey
```

**Review by AVI5HEK — COMMENTED** [review]
The rest LGTM.

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Answer.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package co.nimblehq.avishek.phong.kmmic.domain.model
+
+interface Answerable {
+
```
Comment (suho) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Answer.kt]: Don't need to have a blank line here

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Question.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    TEXTFIELD,
+    DROPDOWN,
+    OUTRO,
+    UNKNOWN;
+
+    companion object {
+        fun fromString(value: String): QuestionDisplayType {
+            return try {
+                QuestionDisplayType.valueOf(value)
+            } catch(exc: IllegalArgumentException) {
```
Comment (suho) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Question.kt]: Please add space between catch and (
```suggestion
            } catch (exc: IllegalArgumentException) {
```

Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Question.kt]: Solved in 38b4965

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Answer.kt (hunk 4 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Answer.kt]: solved in [b24d7ca](https://github.com/nimblehq/avishek-phong-kmm-ic/pull/119/commits/b24d7ca82e18d29a948702230f6b3798cbd7ab08)

File: shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/domain.usecase/GetSurveyDetailUseCaseTest.kt (hunk 3 above)
Comment (phongvhd93) [reply shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/domain.usecase/GetSurveyDetailUseCaseTest.kt]: solved in [05a48e0](https://github.com/nimblehq/avishek-phong-kmm-ic/pull/119/commits/05a48e0eab1776cafd561a4724b5775f2122792c)

File: shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/data.repository/SurveyRepositoryTest.kt (hunk 2 above)
Comment (phongvhd93) [reply shared/src/commonTest/kotlin/co/nimblehq/avishek/phong/kmmic/data.repository/SurveyRepositoryTest.kt]: solved in [05a48e0](https://github.com/nimblehq/avishek-phong-kmm-ic/pull/119/commits/05a48e0eab1776cafd561a4724b5775f2122792c)

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/SurveyRemoteDataSource.kt (hunk 1 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/SurveyRemoteDataSource.kt]: solved in [05a48e0](https://github.com/nimblehq/avishek-phong-kmm-ic/pull/119/commits/05a48e0eab1776cafd561a4724b5775f2122792c)

**Review by AVI5HEK — APPROVED** [review]

**Pull request merged**